R/write_dataset_json.R

```diff
@@ -21,6 +27,47 @@
 write_dataset_json <- function(x, file, pretty=FALSE) {
   stopifnot_datasetjson(x)
 
+  # Find all date, datetime and time columns and convert to character
+  for (y in attr(x,'columns')) {
+
+    # Make sure metadata is compliant
+    if (y$dataType %in% c("date", "datetime", "time") & !("targetDataType" %in% names(y))) {
+      if (!inherits(x[[y$name]], "character")) {
+        stop_write_error(
+          y$name,
+          "If dataType is date, time, or datetime and targetDataType is null, the input variable type must be character"
+        )
+      }
+    }
+
+    if(y$dataType %in% c("date", "datetime", "time") & (!is.null(y$targetDataType) && y$targetDataType == "integer")) {
+      # Convert date
+      if (y$dataType == "date") {
+        x[y$name] <- format(x[[y$name]], "%Y-%m-%d", tz='UTC')
+      }
+
+      # Convert datetime
+      if (y$dataType == "datetime") {
+        # Ensure type and timezone is right.
+        if (!inherits(x[[y$name]], "POSIXlt") || !("UTC" %in% attr(x[[y$name]], 'tzone'))){
+          stop_write_error(y$name, "Date time variable must be provided as POSIXlt type with timezone set to UTC.")
+        }
+        x[y$name] <- strftime(x[[y$name]], "%Y-%m-%dT%H:%M:%S", tz='UTC')
+      }
+
+      # Convert time
+      if (y$dataType == "time") {
+        if (y$dataType == "time" & !inherits(x[[y$name]], "Period")) {
+          stop_write_error(
+            y$name,
+            "If dataType is time and targetDataType is integer, the input variable type must be a lubridate Period object"
```

Comment:
This might be challenging for sponsors.  They would need to know this while developing SDTM and ADaM.  Should we speak to the community here to see how they currently handle numeric times to be sure we are aligned?

Reply:
I see Admiral is using `hms::hms()` and `lubridate::ymd_hms()`

Reply:
So the HMS object we could still convert to character using the exact same code. It's numerically equivalent to the period object. We could extend it to support either a difftime object or a Period object. We could actually also allow for them to use a data.table `ITime` object as well. So that would cover what I'd expect to be the three most common time formats.

Reply:
@nicholas-masel this one is done